module/VuFind/src/VuFind/Controller/ApiKeyController.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+/**
+ * Controller for API keys.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) The National Library of Finland 2025.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Juha Luoma <juha.luoma@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+
+namespace VuFind\Controller;
+
+use VuFind\ApiKey\ApiKeyService;
+use VuFind\Exception\Forbidden;
+
+use function in_array;
+
+/**
+ * Controller for API keys.
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Juha Luoma <juha.luoma@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class ApiKeyController extends AbstractBase
+{
+    /**
+     * Display settings for the API keys
+     *
+     * @return mixed
+     */
+    public function displaySettingsAction()
+    {
+        if (!$user = $this->getUser()) {
+            return $this->forceLogin();
+        }
+        // If not submitted, are we logged in?
+        if (!$this->apiKeysEnabled()) {
+            throw new Forbidden('API keys disabled.');
+        }
+
+        $apiKeyService = $this->getService(ApiKeyService::class);
+        $view = $this->createViewModel();
+        $view->apiKey = $apiKeyService->getApiKeyForUser($user);
+        return $view;
+    }
+
+    /**
+     * Generate an API key for a user.
+     *
+     * @return mixed
+     */
+    public function generateAction()
+    {
+        if (!$user = $this->getUser()) {
+            return $this->forceLogin();
+        }
+        // If not submitted, are we logged in?
+        if (!$this->apiKeysEnabled()) {
+            throw new Forbidden('Access denied.');
+        }
+        if (!$this->permission()->isAuthorized('feature.Developer')) {
+            $this->flashMessenger()->addMessage('Developer::verify_email_address', 'error');
```

Comment:
Do we need an additional check before using this message here? I worry that if the permission is denied for some other reason (e.g. a custom configuration) this very specific message might be confusing/misleading.